Add top 10 tour

rachelekm opened the pull request:
## Overview

Adds Top 10 neighborhood tour:
- Introductory discover modal open on initial page load
    - Selecting "show my recommendations" starts the tour
    - Selecting "list all neighborhoods" skips the tour and updates to list view
- User exits tour by clicking outside of start modal, clicking close button of neighborhood card slide, or clicking neighborhood detail
- Selecting "Show Top 10" button re-opens tour. The start instructions will only display on first tour.

### Checklist

- [ ] `fixup!` commits have been squashed
- [x] `CHANGELOG.md` updated with summary of features or fixes, following
      [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) guidelines
- [ ] `README.md` updated if necessary to reflect the changes
- [ ] Run `./scripts/format` to lint, format, and fix the application source code.
- [ ] CI passes after rebase

### Demo

https://github.com/user-attachments/assets/b2d78e61-79b6-4f03-ba5d-99f56a77952a


## Testing Instructions

 * `scripts/server`
 * Set screen to mobile view and login
 * Confirm start modal appears on discover page
 * Select "list" option and confirm modal closes and discover updates to list view
 * Toggle back to the map and confirm tour is closed and open button appears in top-left corner
 * Click button to open tour
 * Confirm start instructions do not display and tour begins with first neighborhood recommendation
 * Confirm: map bounds update to neighborhood, map styles selected neighborhood bounds, card details are correct, "previous" button is hidden, "details" and "next" buttons appear
 * Click through tour confirm neighborhood cards "previous", "next", and "details" buttons work as expected
 * Confirm on navigating back to discover from neighborhood details that tour is closed
 * Confirm on 10th neighborhood card the "next" button is hidden and closing card ends the tour


Resolves #662

rachelekm committed:
Add discover start tour modal & initial tour logic

rachelekm committed:
Add top 10 tour steps

rachelekm committed:
Add start tour map button

rachelekm commented on src/frontend/src/reducers/userProfile/userSlice.ts:
```diff
@@ -34,17 +35,28 @@ export const userProfileSlice = createSlice({
+            { payload: status }: { payload: boolean }
+        ) => {
+            state.hasViewedStartInstructions = status;
+        },
```
Keeping this status in local storage or in the db felt like overkill (especially since user auth tokens don't expire) so I added it to user profile state for now

rachelekm committed:
Update CHANGELOG

rachelekm commented:
@aaronxsu just rebased this branch with latest develop so this should ready to review now

aaronxsu commented on src/frontend/src/pages/Discover/Map/Map.tsx:
```diff
@@ -192,6 +248,12 @@ const Map = ({ isMobile = true, mapDisplay = true }: Props) => {
                 )}
+                <CustomControlOverlay position="top-left">
+                    <Top10TourButton
+                        isVisible={!!isMobile && !isTop10TourOpen}
```
Wondering if `!!isMobile` could just be `isMobile`. Additionally, should this button be visible also when `mapDisplay && !!topTen.length`? Did I misunderstanding something here?

rachelekm replied:
Good catch! This list of conditions should be switched with the `Top10Tour`'s isOpen condition (which you also caught below)! I fixed this is in 981b0217179f561cf0d75a534ad5900136909b98.

aaronxsu commented on src/frontend/src/pages/Discover/Map/Map.tsx:
```diff
@@ -138,8 +150,52 @@ const Map = ({ isMobile = true, mapDisplay = true }: Props) => {
         <div className={mapContainer()}>
+            <Top10Tour
+                isTop10TourOpen={
+                    !!isMobile &&
```
Same question here. Also, does it have to be this list of conditions? I thought just `isTop10TourOpen` should suffice. Did I miss something here?

rachelekm replied:
Addressed in comment above but good catch, this is fixed in 981b0217179f561cf0d75a534ad5900136909b98

aaronxsu commented on src/frontend/src/components/Top10Tour/StartTourModal.tsx:
```diff
@@ -0,0 +1,75 @@
+import Button from "../base/Button/Button";
+import StartIntro from "assets/icons/start-intro.svg?react";
+import { top10TourStyles } from "./top10Tour.styles";
+import { useTranslation, Trans } from "react-i18next";
```
This is a nitpick- I'd reorder the imports here so that the external dep is at the top and the rest go after it.

rachelekm replied:
Clean up in 37ae782d6500538cb37ca89b88c43a0c8a17d1ea

aaronxsu commented on src/frontend/src/components/Top10Tour/Top10Tour.tsx:
```diff
@@ -0,0 +1,120 @@
+    selectActiveDestination,
+    setHasViewedStartInstructions,
+} from "src/reducers/userProfile/userSlice";
+import { useDispatch } from "react-redux";
```
I'd put this in the top block of imports. Also I'd consider removing the `src/` bit from several imports above.

rachelekm replied:
Cleaned up in 37ae782d6500538cb37ca89b88c43a0c8a17d1ea

aaronxsu commented on src/frontend/src/pages/Discover/Map/Map.tsx:
```diff
@@ -49,6 +57,10 @@ const Map = ({ isMobile = true, mapDisplay = true }: Props) => {
 
+    useEffect(() => {
+        setIsTop10TourOpen(!hasViewedInstructions);
+    }, []);
```
Should `hasViewedInstructions` be a dep?

rachelekm replied:
I was thinking this should run on initial component render, but ended up causing funny side effects. I've adjusted this in 981b0217179f561cf0d75a534ad5900136909b98 so `isTop10TourOpen` and `hasViewedInstructions` are both deps and its more robust

aaronxsu commented on src/frontend/src/components/Top10Tour/Top10Tour.tsx:
```diff
@@ -0,0 +1,120 @@
+    };
+
+    useEffect(() => {
+        if (tourStep < 0) {
```
I'd put 0 and 9 in something like MIN_STEP and MAX_STEP as constants.

rachelekm replied:
Updated in d6166686805672c3f74e43a3e7b42a749568746e

aaronxsu reviewed (commented):
Looking very good and works smoothly! I posted some questions in the comment and would like to know what you think.

Here is something I found when testing, re: map zoom to bound: When I am in the tour, I click the detail button, the detail modal shows up, and then I close the modal, I am back to the map view with tour closed, which is expected according to the instructions. When I click on the button to show the top 10 tour again, the first neighborhood shows up in the card, but the map does not seem to zoom for me. Does it reproduce on your end?

rachelekm committed:
Fix unnecessarily dispatched path/times fetch

rachelekm committed:
Clean up conditional dependencies

rachelekm committed:
Clean up imports

rachelekm commented:
Thanks, @aaronxsu! Just addressed those comments and this is ready for another look! I didn't rebase/address the merge conflicts yet for ease of reviewing (and its just translations).

> Here is something I found when testing, re: map zoom to bound: When I am in the tour, I click the detail button, the detail modal shows up, and then I close the modal, I am back to the map view with tour closed, which is expected according to the instructions. When I click on the button to show the top 10 tour again, the first neighborhood shows up in the card, but the map does not seem to zoom for me. Does it reproduce on your end?

Good catch, I was able to replicate this! Adjusting the `useEffect` and initial top 10 tour open state seemed to do the trick for me, let me know if you're seeing the same!

aaronxsu commented on src/frontend/src/components/Top10Tour/Top10Tour.tsx:
```diff
@@ -0,0 +1,124 @@
+    };
+
+    useEffect(() => {
+        if (tourStep < 0) {
```
I'd replace this with `MIN_STEP` as well.

aaronxsu reviewed (approved):
Looking and working great! This is good to go 🚀

rachelekm commented:
Thanks, @aaronxsu! Will nab that last fix and address the merge conflicts and then get this merged in on successful CI!
